### PR TITLE
Adding remorseful prober

### LINKED ISSUE #387: Tournaments to Reproduce
* http://www.graham-kendall.com/papers/lhk2011.pdf

### PR DESCRIPTION
**Pull request by drvinceknight**
This is the remaining strategy for the list from #387. There is another strategy on that list "contrite tft" but that strategy actually needs to know when a move against it was caused by noise. I suggest (after this) we close #387 and open another issue for that strategy as it will need a bit of a tweak to the `Match` class (or something).

Note that for some of the tests for this strategy I'm not using the helper functions. I didn't seem able to get the required behaviour with them: I got myself a big confused so found it easier to use the verbose calls. Not averse to spending time getting the helper functions working if we are keen to...

### CODE REVIEW

File: axelrod/strategies/prober.py [hunk 1]
```diff
@@ -171,9 +173,61 @@ def strategy(self, opponent):
...
+    Like Naive Prober, but it remembers if the opponent responds to a random
+    defection with a defection by being remorseful and cooperating.
+
+    For reference see: "Engineering Design of Strategies for Winning
+    Iterated Prisoner's Dilemma Competitions" by Jiawei Li, Philip Hingston,
+    and Graham Kendall.  IEEE TRANSACTIONS ON COMPUTATIONAL INTELLIGENCE AND AI
+    IN GAMES, VOL. 3, NO. 4, DECEMBER 2011
+
+    A better description is given in the selfish gene:
+    https://books.google.co.uk/books?id=ekonDAAAQBAJ&pg=PA273&lpg=PA273&dq=remorseful+prober&source=bl&ots=kAeYRYg7GB&sig=RD5-XtDAxzTF9rxRZEWyFjwuKhc&hl=en&sa=X&ved=0ahUKEwiFg--H_qvNAhWXF8AKHQVTAzcQ6AEIKDAC#v=onepage&q=remorseful%20prober&f=false
```
Comment (meatballs) [inline axelrod/strategies/prober.py]: That url can be shortened to just https://books.google.co.uk/books?id=ekonDAAAQBAJ

File: docs/tutorials/advanced/classification_of_strategies.rst [hunk 2]
```diff
@@ -24,7 +24,7 @@ This allows us to, for example, quickly identify all the stochastic
 strategies::
 
     >>> len([s for s in axl.strategies if s().classifier['stochastic']])
-    40
+    41
```
Comment (meatballs) [inline docs/tutorials/advanced/classification_of_strategies.rst]: Doesn't line 33 need to change too?

File: docs/tutorials/advanced/strategy_transformers.rst [hunk 3]
```diff
@@ -104,13 +104,13 @@ The library includes the following transformers:
...
-    >>> import axelrod
-    >>> from axelrod.strategy_transformers import ApologyTransformer
-    >>> ApologizingDefector = ApologyTransformer([D, D], [C, C])(axelrod.Defector)
-    >>> player = ApologizingDefector()
+       >>> import axelrod
+       >>> from axelrod.strategy_transformers import ApologyTransformer
+       >>> ApologizingDefector = ApologyTransformer([D, D], [C, C])(axelrod.Defector)
+       >>> player = ApologizingDefector()
 
 * :code:`DeadlockBreakingTransformer`: Attempts to break :code:`(D, C) -> (C, D)` deadlocks by cooperating::
```
Comment (meatballs) [inline docs/tutorials/advanced/strategy_transformers.rst]: Is this supposed to be here? It doesn't appear to be related to the new strategy.

Comment (drvinceknight) [inline docs/tutorials/advanced/strategy_transformers.rst]: I added this to this PR just to fix the indentation for the docs. It's unrelated.

File: docs/tutorials/advanced/classification_of_strategies.rst (hunk 2 above)
Comment (drvinceknight) [inline docs/tutorials/advanced/classification_of_strategies.rst]: No, this strategy has memory depth 2 as it needs to remember if a 'probe' was punished (so that it can be remorseful).

File: axelrod/strategies/prober.py (hunk 1 above)
Comment (drvinceknight) [inline axelrod/strategies/prober.py]: Good shout, I'll give this a general clean up now :)

File: docs/tutorials/advanced/classification_of_strategies.rst (hunk 2 above)
Comment (meatballs) [inline docs/tutorials/advanced/classification_of_strategies.rst]: Ah yes - I was looking at the memory depth for Naive Prober by mistake

Comment (drvinceknight) [inline docs/tutorials/advanced/classification_of_strategies.rst]: :+1:

File: axelrod/strategies/prober.py [hunk 4]
```diff
@@ -171,9 +173,65 @@ def strategy(self, opponent):
...
+    """
+    Like Naive Prober, but it remembers if the opponent responds to a random
+    defection with a defection by being remorseful and cooperating.
+
+    For reference see: "Engineering Design of Strategies for Winning
+    Iterated Prisoner's Dilemma Competitions" by Jiawei Li, Philip Hingston,
+    and Graham Kendall.  IEEE TRANSACTIONS ON COMPUTATIONAL INTELLIGENCE AND AI
+    IN GAMES, VOL. 3, NO. 4, DECEMBER 2011
+
+    A fuller description is given in "The Selfish Gene"
```
Comment (marcharper) [inline axelrod/strategies/prober.py]: ~~fuller~~ more complete

File: axelrod/tests/unit/test_prober.py [hunk 5]
```diff
@@ -162,3 +161,68 @@ def test_reduction_to_TFT(self):
...
+        opponent.history = [C, D, D]
+        self.assertEqual(player.strategy(opponent), D)
+        self.assertEqual(player.probing, False)
+
+    def test_reduction_to_TFT(self):
+        player = self.player(0)
+        opponent = axelrod.Random()
+        test_responses(self, player, opponent, [C], [C], [C], random_seed=1)
+        test_responses(self, player, opponent, [C], [D], [D])
+        test_responses(self, player, opponent, [C, D], [D, C], [C])
```
Comment (marcharper) [inline axelrod/tests/unit/test_prober.py]: You can use e.g. `attrs = {'probing': False}` with `test_responses` and `responses_test` to check that `self.probing` is appropriately set (also above).

**Comment by marcharper** [comment]
This is a side conversation but you can have a player track its own intended history and then compare to its stored history without modifying the `Match` class. I toyed with a transformer for this a while back, it's a natural use case. Take a look at `TrackHistoryTransformer`.

**Comment by marcharper** [comment]
I had a couple of comments but they are minor so feel free to merge with or without making any changes.

File: axelrod/tests/unit/test_prober.py (hunk 5 above)
Comment (drvinceknight) [inline axelrod/tests/unit/test_prober.py]: Yeah I tried the various helper functions but with no success and was getting frustrated as to whether I was making mistakes with the strategy or misusing the helper functions so it just felt easier to use the verbose calls...

I'll have another go. It might have just been me getting myself confused.

**Comment by drvinceknight** [comment]
> This is a side conversation but you can have a player track its own intended history and then compare to its stored history without modifying the Match class. I toyed with a transformer for this a while back, it's a natural use case. Take a look at TrackHistoryTransformer.

So this would be helpful for contrite TFT right?

File: axelrod/strategies/prober.py (hunk 4 above)
Comment (drvinceknight) [inline axelrod/strategies/prober.py]: Yup.

**Comment by drvinceknight** [comment]
> I had a couple of comments but they are minor so feel free to merge with or without making any changes.

Don't merge yet: will give the test helper functions a better go.

**Comment by marcharper** [comment]
> So this would be helpful for contrite TFT right?
I presume so!

**Comment by drvinceknight** [comment]
Am I missing something here:

```
  200         #random.seed(0)
  201         #player.history = [C]
  202         #opponent.history = [C]
  203         #self.assertEqual(player.strategy(opponent), D)  # Random defection
  204         #self.assertEqual(player.probing, True)
  205         test_responses(self, player, opponent, [C], [C], [D], random_seed=0,
  206                        attrs={'probing': False}
  207                        )
```

Commenting out lines 200 -> 204 and replacing with 205 -> 205 (as above) do not give the same result:

```
Axelrod(387) ✗: python3.5 -m unittest axelrod.tests.unit.test_prober
/Users/vince/src/Axelrod/axelrod/tests/unit/test_player.py:175: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  argspec = inspect.getargspec(PlayerClass.__init__)
..............................................F....
======================================================================
FAIL: test_remorse (axelrod.tests.unit.test_prober.TestRemorsefulProber)
After probing, if opponent retaliates, will offer a C
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vince/src/Axelrod/axelrod/tests/unit/test_prober.py", line 206, in test_remorse
    attrs={'probing': False}
  File "/Users/vince/src/Axelrod/axelrod/tests/unit/test_player.py", line 100, in test_responses
    test_class.assertEqual(getattr(P1, attr), value)
AssertionError: True != False

----------------------------------------------------------------------
Ran 51 tests in 0.044s

FAILED (failures=1)
```

Looking at the source code it certainly should be, any ideas?

**Comment by drvinceknight** [comment]
Ignore me. That was a stupid question (I was misreading).

**Comment by drvinceknight** [comment]
**But** this diff should give the same behaviour right? 

```
-        random.seed(0)
-        player.history = [C]
-        opponent.history = [C]
-        self.assertEqual(player.strategy(opponent), D)  # Random defection
-        self.assertEqual(player.probing, True)
+        test_responses(self, player, opponent, [C], [C], [D], random_seed=0,
+                       attrs={'probing': True})

-        player.history = [C, D]
-        opponent.history = [C, D]
-        self.assertEqual(player.strategy(opponent), C)  # Remorse
-        self.assertEqual(player.probing, False)
+        test_responses(self, player, opponent, [C, D], [C, D], [C],
+                       attrs={'probing': False})

-        player.history = [C, D, C]
-        opponent.history = [C, D, D]
-        self.assertEqual(player.strategy(opponent), D)
-        self.assertEqual(player.probing, False)
+        test_responses(self, player, opponent, [C, D, C], [C, D, D], [D],
+                       attrs={'probing': False})
```

It does not: the second test does not show remorse. I think the helper functions are doing something in the background that I don't understand (and becoming less of a fan of...). Am I just missing something and getting myself confused?

**Comment by drvinceknight** [comment]
I've pushed 27a065e with the above diff. The tests fail.

**Comment by drvinceknight** [comment]
ba1cce5 reverts the commit that fails. I'll leave it there in case it's helpful and/or an issue needs to be raised (does something need to be better documented?) but I'm happy (if everyone else is) for this to be merged as is. Not using the helper functions is my preference (I think): it's verbose and explicit.

**Comment by marcharper** [comment]
Your tests are fine. I'll take a closer look at the `attrs` options, they've worked for several other strategies.
